File: src/main/java/algo/backtrack/Combinations_77.java
package algo.backtrack;

import java.util.ArrayList;
import java.util.List;

public class Combinations_77 {
    public static List<List<Integer>> combine(int n, int k) {
        var combinations = new ArrayList<List<Integer>>();
        backtrack(n, k, 1, new ArrayList<>(), combinations);
        return combinations;
    }

    private static void backtrack(int n, int k, int start, List<Integer> path,
                                  List<List<Integer>> combinations) {
        if (k == 0) {
            combinations.add(new ArrayList<>(path));
            return;
        }
        for (var i = start; i <= n; i++) {
            path.add(i);
            backtrack(n, k - 1, i + 1, path, combinations);
            path.remove(path.size() - 1);
        }
    }

    public static void main(String[] args) {
        var combinations = combine(4, 2);
        combinations.forEach(System.out::println);
    }
}

File: src/main/kotlin/algo/backtrack/Combinations_77.kt
package algo.backtrack

class KtCombinations_77 {
    fun combine(n: Int, k: Int): List<List<Int>> {
        fun backtrack(n: Int, k: Int, start: Int, all: MutableList<List<Int>>, combination: MutableList<Int>) {
            if (k == 0) {
                all.add(combination.toList())
            }
            for (i in start..n) {
                combination.add(i)
                backtrack(n, k - 1, i + 1, all, combination)
                combination.removeLast()
            }
        }

        val combinations: MutableList<List<Int>> = mutableListOf()
        backtrack(n, k, 1, combinations, mutableListOf())
        return combinations
    }
}

fun main() {
    val combinations = KtCombinations_77()
    println(combinations.combine(4, 2))
}